### PR TITLE
Feature/response validation status code based

### PR DESCRIPTION
**Pull request by allevo**
PR for #84

- Keep the back compatibility for out schema definition
- Add response key in schema definition that has an higher priority over out definition
- Fallbacks implemented (200 -> 2xx if 200 schema is not defined)
- Some test: to do better

### CODE REVIEW

File: lib/validation.js [hunk 1]
```diff
@@ -61,8 +80,25 @@ function validate (handle, params, body, query) {
...
+    return responseSchema[statusCode](data)
+  }
+  var fallbackStatusCode = (statusCode + '')[0] + 'xx'
+  if (responseSchema[fallbackStatusCode]) {
+    return responseSchema[fallbackStatusCode](data)
+  }
+}
+
+function serialize (handle, data, statusCode) {
+  let output
```
Comment (mcollina) [inline lib/validation.js]: please use var for consistency

File: test/output-validation.test.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
+'use strict'
+
+const t = require('tap')
+const test = t.test
+const request = require('request')
+const fastify = require('..')()
+
+const schema = {
+  response: {
+    '200': {
```
Comment (mcollina) [inline test/output-validation.test.js]: you should be able to remove the '

Comment (allevo) [inline test/output-validation.test.js]: Yes, if the key is a number I'm able to use it directly. But sadly, with the 2xx the ' is required

File: lib/validation.js [hunk 3]
```diff
@@ -9,10 +9,25 @@ const bodySchema = Symbol('body-schema')
...
 
 const schemas = require('./schemas.json')
 const inputSchemaError = fastJsonStringify(schemas.inputSchemaError)
 
+function getValidatorForStatusCodeSchema (statusCodeDefinition) {
+  return fastJsonStringify(statusCodeDefinition)
+}
+
+function getResponseSchema (responseSchemaDefinition) {
+  var schemas = { }
```
Comment (mcollina) [inline lib/validation.js]: I prefer empty literals to be `{}`, not  `{ }`.

File: lib/validation.js [hunk 4]
```diff
@@ -9,10 +9,25 @@ const bodySchema = Symbol('body-schema')
...
+  var statusCodes = Object.keys(responseSchemaDefinition)
+  for (var i = 0; i < statusCodes.length; i++) {
+    var statusCode = statusCodes[i]
+    schemas[statusCode] = getValidatorForStatusCodeSchema(responseSchemaDefinition[statusCode])
+  }
```
Comment (mcollina) [inline lib/validation.js]: This should be on a non-hot path, i.e. it is not in the rendering pipeline. If that's the case you can use a `forEach`, it's much more readable.

Comment (allevo) [inline lib/validation.js]: Can I use reduce array method over `statusCodes`?

Comment (mcollina) [inline lib/validation.js]: Yes.

**Comment by allevo** [comment]
@mcollina  how can i run the benchmaks?
Is there a script for running those test?

**Comment by delvedor** [comment]
> @mcollina how can i run the benchmarks?
Is there a script for running those test?

We must update the docs, the benchmark are taken by running [this](https://github.com/fastify/fastify/blob/master/examples/example.js) file, with autocannon:
```
$ autocannon -c 100 -d 5 -p 10 localhost:3000
```

To check if the new code introduces some regression you must bench the *master* branch and then the branch with your changes :)

File: lib/validation.js (hunk 1 above)
Comment (allevo) [inline lib/validation.js]: done

File: test/output-validation.test.js (hunk 2 above)
Comment (allevo) [inline test/output-validation.test.js]: done

File: lib/validation.js (hunk 3 above)
Comment (allevo) [inline lib/validation.js]: done

File: lib/validation.js (hunk 4 above)
Comment (allevo) [inline lib/validation.js]: done

**Comment by allevo** [comment]
@delvedor thanks!
@mcollina the results:

## feature/response-validation-status-code-based (97edbec)

    $ autocannon -c 100 -d 5 -p 10 localhost:3000
    Running 5s test @ http://localhost:3000
    100 connections with 10 pipelining factor
    
    Stat         Avg      Stdev     Max    
    Latency (ms) 9.31     18.29     187    
    Req/Sec      10343.21 901.95    11375  
    Bytes/Sec    1.53 MB  139.95 kB 1.7 MB 
    
    52k requests in 5s, 7.71 MB read
    
    $ autocannon -c 100 -d 5 -p 10 localhost:3000
    Running 5s test @ http://localhost:3000
    100 connections with 10 pipelining factor
    
    Stat         Avg     Stdev    Max    
    Latency (ms) 8.96    15.34    146    
    Req/Sec      10759.2 456.93   11415  
    Bytes/Sec    1.61 MB 71.79 kB 1.7 MB 
    
    54k requests in 5s, 8.02 MB read
    
## master (4f62c4b)

    $ autocannon -c 100 -d 5 -p 10 localhost:3000
    Running 5s test @ http://localhost:3000
    100 connections with 10 pipelining factor
    
    Stat         Avg      Stdev     Max    
    Latency (ms) 9.38     18.95     198    
    Req/Sec      10303.21 894.64    11215  
    Bytes/Sec    1.53 MB  145.96 kB 1.7 MB 
    
    52k requests in 5s, 7.68 MB read
    
    $ autocannon -c 100 -d 5 -p 10 localhost:3000
    Running 5s test @ http://localhost:3000
    100 connections with 10 pipelining factor
    
    Stat         Avg     Stdev    Max    
    Latency (ms) 9.24    16.71    162    
    Req/Sec      10429.6 509.13   11143  
    Bytes/Sec    1.55 MB 76.43 kB 1.7 MB 
    
    52k requests in 5s, 7.77 MB read


Almost the same.
The code adds two `if` at runtime, so the benchmarks are not affected.

**Comment by allevo** [comment]
For memory efficiency, I can use `Object.create(null)` instead of `{}` in the reduce method.

**Comment by mcollina** [comment]
Where?

File: lib/validation.js [hunk 5]
```diff
@@ -7,23 +7,36 @@ const ajv = new Ajv({ coerceTypes: true })
...
+function getValidatorForStatusCodeSchema (statusCodeDefinition) {
+  return fastJsonStringify(statusCodeDefinition)
+}
+
+function getResponseSchema (responseSchemaDefinition) {
+  var statusCodes = Object.keys(responseSchemaDefinition)
+  return statusCodes.reduce(function (r, statusCode) {
+    r[statusCode] = getValidatorForStatusCodeSchema(responseSchemaDefinition[statusCode])
+    return r
+  }, {})
```
Comment (allevo) [inline lib/validation.js]: @mcollina  here

Comment (mcollina) [inline lib/validation.js]: there are not that many of those objects, there is no need.

File: lib/validation.js [hunk 6]
```diff
@@ -61,8 +74,19 @@ function validate (handle, params, body, query) {
...
-  return handle[outputSchema](data)
+function serializeAccordingToResponseSchema (responseSchema, data, statusCode) {
+  if (responseSchema[statusCode]) {
+    return responseSchema[statusCode](data)
+  }
+  var fallbackStatusCode = (statusCode + '')[0] + 'xx'
+  if (responseSchema[fallbackStatusCode]) {
+    return responseSchema[fallbackStatusCode](data)
+  }
+  return responseSchema.default(data)
```
Comment (delvedor) [inline lib/validation.js]: Probably store `fast-safe-strinfgify` inside `default` is not needed, we can just call `fastSafeStringify` directly here. @mcollina thoughts?

Comment (mcollina) [inline lib/validation.js]: No there is no need.

File: lib/validation.js [hunk 7]
```diff
@@ -61,8 +74,19 @@ function validate (handle, params, body, query) {
...
+}
+
+function serialize (handle, data, statusCode) {
+  return serializeAccordingToResponseSchema(handle[responseSchema], data, statusCode)
 }
```
Comment (mcollina) [inline lib/validation.js]: can you remove this function call completely, and rename `serializeAccordingToResponseSchema`  to `serialize`?

Comment (allevo) [inline lib/validation.js]: done

File: lib/validation.js [hunk 8]
```diff
@@ -7,23 +7,36 @@ const ajv = new Ajv({ coerceTypes: true })
...
+  return statusCodes.reduce(function (r, statusCode) {
+    r[statusCode] = getValidatorForStatusCodeSchema(responseSchemaDefinition[statusCode])
+    return r
+  }, {})
+}
+
 function build (opts) {
   if (!opts.schema) {
-    opts[outputSchema] = fastSafeStringify
+    opts[responseSchema] = {default: fastSafeStringify}
```
Comment (mcollina) [inline lib/validation.js]: there is no need to configure a default for `fastSafeStringify` use in place of the `out[responseSchema].default()` call.

**Comment by mcollina** [comment]
@allevo ping :)

File: lib/validation.js (hunk 8 above)
Comment (allevo) [inline lib/validation.js]: Done

File: lib/validation.js [hunk 9]
```diff
@@ -7,23 +7,36 @@ const ajv = new Ajv({ coerceTypes: true })
...
 
-  if (opts.schema.out) {
-    opts[outputSchema] = fastJsonStringify(opts.schema.out)
+  if (opts.schema.response) {
+    opts[responseSchema] = getResponseSchema(opts.schema.response)
   } else {
-    opts[outputSchema] = fastSafeStringify
+    opts[responseSchema] = {}
   }
+  opts[responseSchema].default = fastSafeStringify
```
Comment (allevo) [inline lib/validation.js]: Done

File: test/internals/validation.test.js [hunk 10]
```diff
@@ -49,7 +49,7 @@ test('build schema - output schema', t => {
     }
   }
   validation.build(opts)
-  t.is(typeof opts[symbols.responseSchema].default, 'function')
+  t.is(typeof opts[symbols.responseSchema].default, 'undefined')
```
Comment (delvedor) [inline test/internals/validation.test.js]: This can be removed :)

**Comment by delvedor** [comment]
@mcollina if everything is ok to you I can proceed with merge and release :)

**Comment by mcollina** [comment]
yes go ahead
Il giorno gio 1 giu 2017 alle 13:12 Tomas Della Vedova <
notifications@github.com> ha scritto:

**Comment by delvedor** [comment]
@allevo would you like to join our team and help us maintaining Fastify? :)

**Comment by allevo** [comment]
@delvedor yes and thanks!!

**Pull request merged**